src/dk/simongray/datalinguist/ml/crf.clj

```diff
@@ -0,0 +1,117 @@
+   :maxLeft           "1"
+   ;; these are the features we'd like to train with
+   ;; some are discussed below , the rest can be
+   ;; understood by looking at NERFeatureFactory
+   :useClassFeature   "true"
+   :useWord           "true"
+   ;; word character ngrams will be included up to length 6 as prefixes
+   ;; and suffixes only
+   :useNGrams         "true"
+   :noMidNGrams       "true"
+   :maxNGramLeng      "6"
+   :usePrev           "true"
+   :useNext           "true"
+   :useDisjunctive    "true"
+   :useSequences      "true"
+   :usePrevSequences  "true"
+   ;; the last 4 properties deal with word shape features
+   :useTypeSeqs       "true"
+   :useTypeSeqs2      "true"
+   :useTypeySequences "true"
+   ;;wordShape chris2useLC
+   :wordShape         "none"
+   #_#_:useBoundarySequences "true"
+   #_#_:useNeighborNGrams "true"
+   #_#_:useTaggySequences "true"
+   #_#_:printFeatures "true"
+   #_#_:saveFeatureIndexToDisk "true"
+   #_#_:useObservedSequencesOnly "true"
+   #_#_:useWordPairs "true"})
+
+(def file-opts
+  {:map "word=0,answer=1"})
+
+(defn- train-crf
+  [dataset word-col tag-col opts]
+  (let [training-file (File/createTempFile "ner-data" ".tsv")
+        props         (-> opts
+                          (merge file-opts)
+                          (assoc
+                            "trainFile" (.getAbsolutePath training-file))
+                          (util/properties))]
+    (-> dataset
+        (ds/select-columns [word-col tag-col])
+        (ds/write! (.getAbsolutePath training-file) {:skip-headers true}))
+    (doto (CRFClassifier. props)
+      (.train))))
+
+(defn- triple->seq
+  [triples]
+  (map #(.asList %) triples))
+
+(defn- predict-crf
+  [dataset word-col model]
+  (->> (get dataset word-col)
+       (map #(triple->seq (.classifyToCharacterOffsets model %)))
+       (ds/new-column :ner)
+       (ds/add-column dataset)))
+
+(defn train
+  "Train an edu.stanford.nlp.ie.crf.CRFClassifier.
+
+  The training data needs to be given as tokens in the feature column of the
+  dataset and the gold class needs to be given in the inference-target column of
+  the dataset. The options maps can contain any option specified in the javadoc
+  of CRFClassifier.
+
+  This methods gets called indirectly by `tech.ml`."
+  [feature-dataset target-dataset opts]
+  (train-crf (ds/concat feature-dataset target-dataset)
+             (first (ds/column-names feature-dataset))
+             (first (ds/column-names target-dataset))
+             (merge default-opts (:crf-options opts))))
+
+;; TODO: what's with the unused model param?
+(defn predict
+  "Predict on new data using the CRF classifier.
+
+  The data need to be given a texts in the feature column of the dataset.
+
+  This method gets called indirectly by `tech.ml`."
+  [feature-dataset thawed-model model]
+  (let [word-col (first (ds/column-names feature-dataset))]
+    (predict-crf feature-dataset word-col thawed-model)))
+
+(defn define-model!
```

Comment:
I think this is not in the spirit of the `tech.ml` architecture.
The "model registration` should happen ones.

But indeed we need to be able to change the configuration, but at runtime by calling `train` and pass options.
This happens in line 89, the :crf-options passed in will overwrite the default options.

This need to be documented, indeed.

'Tech.ml' has a way to document all options a model can take as a "map".
This is shown here:
https://github.com/scicloj/scicloj.ml.smile/blob/49a753c8ce6ae991b591a563f6a723c660ddb776/src/scicloj/ml/smile/sparse_logreg.clj#L48

I was planning to do this as well.

This metadata can then be used to generate documentation:
https://scicloj.github.io/scicloj.ml/userguide-models.html

Reply:
I see. I guess I should revert this to just {} at the moment, remove the `define-model!` fn and leave in a TODO for later?

Reply:
yes, just revert it.

I can add all parameters know as well, before you merge.

Reply:
yes